File: tictactoe.js
/* ============================================================
 * tictactoe.js
 * Frontend controller for the Tic-Tac-Toe widget.
 *
 * All game rules live in tictactoe.c, compiled to tictactoe.wasm.
 * This file's only job is: load that module, translate DOM events
 * into calls on it, and render whatever it reports back. No game
 * logic (win checks, turn order, AI) is duplicated here.
 * ============================================================ */

(function () {
  "use strict";

  const MODE_PVP = 0;
  const MODE_PVC = 1;

  const STATE_ONGOING = 0;
  const STATE_X_WINS = 1;
  const STATE_O_WINS = 2;
  const STATE_DRAW = 3;

  const WIN_LINES = [
    [0, 1, 2], [3, 4, 5], [6, 7, 8],
    [0, 3, 6], [1, 4, 7], [2, 5, 8],
    [0, 4, 8], [2, 4, 6],
  ];

  const WASM_PATH = new URL("tictactoe.wasm", document.currentScript.src);

  const state = {
    exports: null,
    mode: MODE_PVC,
    scores: { x: 0, o: 0, draws: 0 },
    locked: false, // true while the computer is "thinking" or game is over
  };

  let root, boardEl, cells, statusEl, restartBtn, modeButtons, scoreEl;

  function markSvg(mark, animate) {
    const drawClass = animate ? " ttt-mark-draw-on" : "";
    if (mark === 1) {
      return (
        `<svg viewBox="0 0 100 100" class="ttt-mark-x${drawClass}" aria-hidden="true">` +
        `<path class="ttt-mark-path" pathLength="1" d="M22 22 L78 78"></path>` +
        `<path class="ttt-mark-path" pathLength="1" style="animation-delay:0.12s" d="M78 22 L22 78"></path>` +
        `</svg>`
      );
    }
    if (mark === 2) {
      return (
        `<svg viewBox="0 0 100 100" class="ttt-mark-o${drawClass}" aria-hidden="true">` +
        `<circle class="ttt-mark-path" pathLength="1" cx="50" cy="50" r="32"></circle>` +
        `</svg>`
      );
    }
    return "";
  }

  function findWinningLine(board) {
    for (const line of WIN_LINES) {
      const [a, b, c] = line;
      if (board[a] !== 0 && board[a] === board[b] && board[b] === board[c]) {
        return line;
      }
    }
    return null;
  }

  function readBoard() {
    const board = new Array(9);
    for (let i = 0; i < 9; i++) board[i] = state.exports.get_cell(i);
    return board;
  }

  function render(justPlayedIndex) {
    const e = state.exports;
    const board = readBoard();
    const gameState = e.get_game_state();
    const currentPlayer = e.get_current_player();
    const winLine = gameState === STATE_X_WINS || gameState === STATE_O_WINS
      ? findWinningLine(board)
      : null;

    board.forEach((mark, i) => {
      const cell = cells[i];
      const animate = i === justPlayedIndex;
      cell.innerHTML = markSvg(mark, animate);
      cell.disabled = mark !== 0 || gameState !== STATE_ONGOING;
      cell.classList.toggle("is-win", !!winLine && winLine.includes(i));
      cell.setAttribute(
        "aria-label",
        mark === 0 ? `Empty cell ${i + 1}` : `Cell ${i + 1}: ${mark === 1 ? "X" : "O"}`
      );
    });

    renderStatus(gameState, currentPlayer);
    renderScore();

    state.locked = gameState !== STATE_ONGOING;
  }

  function renderStatus(gameState, currentPlayer) {
    statusEl.classList.remove("is-win-you", "is-win-x", "is-win-o", "is-lose", "is-draw");
    let html = "";
    let text = "";

    if (gameState === STATE_ONGOING) {
      const mark = currentPlayer; // 1 = X, 2 = O
      html = markSvg(mark, false);
      if (state.mode === MODE_PVC) {
        text = currentPlayer === 1 ? "Your move" : "Computer is thinking\u2026";
      } else {
        text = `Player ${currentPlayer === 1 ? "X" : "O"}\u2019s move`;
      }
    } else if (gameState === STATE_X_WINS) {
      if (state.mode === MODE_PVC) {
        text = "You Win";
        statusEl.classList.add("is-win-you");
      } else {
        text = "Player X Wins!";
        statusEl.classList.add("is-win-x");
      }
    } else if (gameState === STATE_O_WINS) {
      if (state.mode === MODE_PVC) {
        text = "You Lose";
        statusEl.classList.add("is-lose");
      } else {
        text = "Player O Wins!";
        statusEl.classList.add("is-win-o");
      }
    } else if (gameState === STATE_DRAW) {
      text = "Draw Game";
      statusEl.classList.add("is-draw");
    }

    statusEl.innerHTML = `<span class="ttt-turn-mark">${html}</span><span>${text}</span>`;
  }

  function renderScore() {
    const leftLabel = state.mode === MODE_PVC ? "You" : "X";
    const rightLabel = state.mode === MODE_PVC ? "CPU" : "O";
    scoreEl.innerHTML =
      `${leftLabel} <span>${state.scores.x}</span> &nbsp;\u2014&nbsp; ` +
      `Draws <span>${state.scores.draws}</span> &nbsp;\u2014&nbsp; ` +
      `${rightLabel} <span>${state.scores.o}</span>`;
  }

  function tallyIfFinished(gameState) {
    if (gameState === STATE_X_WINS) state.scores.x++;
    else if (gameState === STATE_O_WINS) state.scores.o++;
    else if (gameState === STATE_DRAW) state.scores.draws++;
  }

  function maybeRunComputer() {
    const e = state.exports;
    if (
      state.mode === MODE_PVC &&
      e.get_game_state() === STATE_ONGOING &&
      e.get_current_player() === 2
    ) {
      state.locked = true;
      setTimeout(() => {
        const played = e.computer_move();
        render(played);
        tallyIfFinished(e.get_game_state());
        renderScore();
      }, 380); // brief pause so the move doesn't feel instant/robotic
    }
  }

  function onCellClick(evt) {
    if (state.locked) return;
    const index = Number(evt.currentTarget.dataset.index);
    const e = state.exports;

    if (state.mode === MODE_PVC && e.get_current_player() !== 1) return;

    const result = e.make_move(index);
    if (result < 0) return; // invalid move (shouldn't happen: cell already disabled)

    render(index);
    tallyIfFinished(result);
    maybeRunComputer();
  }

  function startNewGame(mode) {
    state.mode = mode;
    state.exports.reset_game(mode);
    modeButtons.forEach((btn) =>
      btn.classList.toggle("is-active", Number(btn.dataset.mode) === mode)
    );
    render(-1);
  }

  function attachEvents() {
    cells.forEach((cell) => cell.addEventListener("click", onCellClick));
    restartBtn.addEventListener("click", () => startNewGame(state.mode));
    modeButtons.forEach((btn) =>
      btn.addEventListener("click", () => startNewGame(Number(btn.dataset.mode)))
    );
  }

  async function loadWasm(url) {
    if (typeof WebAssembly.instantiateStreaming === "function") {
      try {
        return await WebAssembly.instantiateStreaming(fetch(url), {});
      } catch (err) {
        // Falls through to the ArrayBuffer path below (e.g. the server
        // didn't send a application/wasm content-type).
      }
    }
    const resp = await fetch(url);
    const bytes = await resp.arrayBuffer();
    return WebAssembly.instantiate(bytes, {});
  }

  async function init() {
    root = document.getElementById("ttt-widget");
    if (!root) return;

    boardEl = root.querySelector(".ttt-board");
    cells = Array.from(root.querySelectorAll(".ttt-cell"));
    statusEl = root.querySelector(".ttt-status");
    restartBtn = root.querySelector(".ttt-restart");
    modeButtons = Array.from(root.querySelectorAll(".ttt-mode button"));
    scoreEl = root.querySelector(".ttt-score");

    const { instance } = await loadWasm(WASM_PATH);

    state.exports = instance.exports;
    state.exports.seed_rng((Date.now() & 0xffffffff) >>> 0);

    attachEvents();
    startNewGame(state.mode);
  }

  if (document.readyState === "loading") {
    document.addEventListener("DOMContentLoaded", init);
  } else {
    init();
  }
})();

File: tictactoe.c
/* ============================================================
 * tictactoe.c  --  Tic-Tac-Toe game logic
 *
 * This is the "backend" of the game: all rules, turn handling,
 * win/draw detection and the computer opponent live here, in C.
 * It is compiled straight to WebAssembly (no Emscripten runtime,
 * no libc) so the browser can call these functions directly.
 *
 * Build (see build.sh):
 *   clang --target=wasm32 -O2 -nostdlib -fno-builtin \
 *         -Wl,--no-entry -Wl,--strip-all \
 *         -o tictactoe.wasm tictactoe.c
 *
 * Cell values on the board:
 *   0 = empty, 1 = X, 2 = O
 *
 * Game state values:
 *   0 = in progress, 1 = X wins, 2 = O wins, 3 = draw
 * ============================================================ */

#define EMPTY 0
#define PLAYER_X 1
#define PLAYER_O 2

#define STATE_ONGOING 0
#define STATE_X_WINS  1
#define STATE_O_WINS  2
#define STATE_DRAW    3

#define MODE_PVP 0
#define MODE_PVC 1

static unsigned char board[9];
static int current_player;   /* whose turn it is: PLAYER_X or PLAYER_O   */
static int game_mode;        /* MODE_PVP or MODE_PVC                     */
static int game_state;       /* STATE_ONGOING / _X_WINS / _O_WINS / DRAW */
static unsigned int rng_state = 2463534242u; /* xorshift32 seed */

/* All 8 possible winning lines */
static const unsigned char LINES[8][3] = {
    {0, 1, 2}, {3, 4, 5}, {6, 7, 8}, /* rows    */
    {0, 3, 6}, {1, 4, 7}, {2, 5, 8}, /* columns */
    {0, 4, 8}, {2, 4, 6}             /* diagonals */
};

/* ---- tiny xorshift32 PRNG so we don't need libc's rand() ---- */
static unsigned int next_random(void) {
    unsigned int x = rng_state;
    x ^= x << 13;
    x ^= x >> 17;
    x ^= x << 5;
    rng_state = x;
    return x;
}

__attribute__((export_name("seed_rng")))
void seed_rng(unsigned int seed) {
    rng_state = seed ? seed : 1u;
}

/* Scan the board and return the current game state without mutating it */
static int evaluate_board(void) {
    for (int i = 0; i < 8; i++) {
        unsigned char a = board[LINES[i][0]];
        unsigned char b = board[LINES[i][1]];
        unsigned char c = board[LINES[i][2]];
        if (a != EMPTY && a == b && b == c) {
            return (a == PLAYER_X) ? STATE_X_WINS : STATE_O_WINS;
        }
    }
    for (int i = 0; i < 9; i++) {
        if (board[i] == EMPTY) return STATE_ONGOING;
    }
    return STATE_DRAW;
}

__attribute__((export_name("reset_game")))
void reset_game(int mode) {
    for (int i = 0; i < 9; i++) board[i] = EMPTY;
    current_player = PLAYER_X; /* X always opens */
    game_mode = mode;
    game_state = STATE_ONGOING;
}

__attribute__((export_name("get_cell")))
int get_cell(int index) {
    if (index < 0 || index > 8) return -1;
    return board[index];
}

__attribute__((export_name("get_current_player")))
int get_current_player(void) {
    return current_player;
}

__attribute__((export_name("get_game_state")))
int get_game_state(void) {
    return game_state;
}

__attribute__((export_name("get_game_mode")))
int get_game_mode(void) {
    return game_mode;
}

/* Places the current player's mark on `cell`.
 * Returns the resulting game_state, or a negative error code:
 *   -1 game already over, -2 out of range, -3 cell occupied
 */
__attribute__((export_name("make_move")))
int make_move(int cell) {
    if (game_state != STATE_ONGOING) return -1;
    if (cell < 0 || cell > 8) return -2;
    if (board[cell] != EMPTY) return -3;

    board[cell] = (unsigned char)current_player;
    game_state = evaluate_board();
    if (game_state == STATE_ONGOING) {
        current_player = (current_player == PLAYER_X) ? PLAYER_O : PLAYER_X;
    }
    return game_state;
}

/* Would placing `player` at `cell` win the game? (pure test, no mutation) */
static int is_winning_move(int cell, int player) {
    unsigned char saved = board[cell];
    board[cell] = (unsigned char)player;
    int result = evaluate_board();
    board[cell] = saved;
    return (player == PLAYER_X) ? (result == STATE_X_WINS)
                                 : (result == STATE_O_WINS);
}

/* Computer always plays as O.
 * Strategy: win if possible -> block opponent's win -> take center ->
 *           take a corner -> take an edge. Ties broken randomly.
 * Returns the cell it played, or -1 if it wasn't the computer's turn.
 */
__attribute__((export_name("computer_move")))
int computer_move(void) {
    if (game_state != STATE_ONGOING) return -1;
    if (game_mode != MODE_PVC || current_player != PLAYER_O) return -1;

    int choice = -1;

    /* 1. Take a winning move if one exists */
    for (int i = 0; i < 9 && choice < 0; i++) {
        if (board[i] == EMPTY && is_winning_move(i, PLAYER_O)) choice = i;
    }

    /* 2. Otherwise block the opponent's winning move */
    if (choice < 0) {
        for (int i = 0; i < 9 && choice < 0; i++) {
            if (board[i] == EMPTY && is_winning_move(i, PLAYER_X)) choice = i;
        }
    }

    /* 3. Otherwise take the center */
    if (choice < 0 && board[4] == EMPTY) choice = 4;

    /* 4. Otherwise take a random open corner */
    if (choice < 0) {
        int corners[4] = {0, 2, 6, 8};
        int open[4], count = 0;
        for (int i = 0; i < 4; i++) {
            if (board[corners[i]] == EMPTY) open[count++] = corners[i];
        }
        if (count > 0) choice = open[next_random() % (unsigned int)count];
    }

    /* 5. Otherwise take a random open edge */
    if (choice < 0) {
        int edges[4] = {1, 3, 5, 7};
        int open[4], count = 0;
        for (int i = 0; i < 4; i++) {
            if (board[edges[i]] == EMPTY) open[count++] = edges[i];
        }
        if (count > 0) choice = open[next_random() % (unsigned int)count];
    }

    if (choice < 0) return -1; /* board full, shouldn't happen if state was ongoing */

    board[choice] = PLAYER_O;
    game_state = evaluate_board();
    if (game_state == STATE_ONGOING) current_player = PLAYER_X;
    return choice;
}
